[CI] Various tests in SearchCancellationIT failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/a34zi62vvrimk/tests/:server:internalClusterTest/org.elasticsearch.search.SearchCancellationIT/testCancelFailedSearchWhenPartialResultDisallowed
**Reproduction line:**
```
./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.search.SearchCancellationIT.testCancelFailedSearchWhenPartialResultDisallowed" -Dtests.seed=C0942422CB06E4E8 -Dtests.locale=ja-JP -Dtests.timezone=Asia/Samarkand -Druntime.java=17 -Dtests.fips.enabled=true
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.search.SearchCancellationIT&tests.test=testCancelFailedSearchWhenPartialResultDisallowed


**Failure excerpt:**
```
java.lang.AssertionError: The Coordinator should have one SearchTask.
Expected: a collection with size <1>
     but: collection size was <0>

  at __randomizedtesting.SeedInfo.seed([C0942422CB06E4E8:8A88C92C7B4503D0]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.elasticsearch.search.SearchCancellationIT.lambda$testCancelFailedSearchWhenPartialResultDisallowed$4(SearchCancellationIT.java:283)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1237)
  at org.elasticsearch.search.SearchCancellationIT.testCancelFailedSearchWhenPartialResultDisallowed(SearchCancellationIT.java:281)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:833)

```

This test suite fails a few times a day and has for some time now. The failure exception seems fairly evenly split between the one linked in the original description of this issue and [this one here](https://gradle-enterprise.elastic.co/s/76fkjfpj3jnn4/tests/task/:server:internalClusterTest/details/org.elasticsearch.search.SearchCancellationIT/testCancellationOfScrollSearches?top-execution=1). Not sure if they are the same root cause or not.

```
java.lang.AssertionError: 	
Expected: a value greater than <0>	
     but: <0> was equal to <0>	
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)	
	at org.junit.Assert.assertThat(Assert.java:956)	
	at org.junit.Assert.assertThat(Assert.java:923)	
	at org.elasticsearch.test.AbstractSearchCancellationTestCase.lambda$awaitForBlock$0(AbstractSearchCancellationTestCase.java:103)
```

--- a/server/src/internalClusterTest/java/org/elasticsearch/search/SearchCancellationIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/search/SearchCancellationIT.java
@@ -8,7 +8,6 @@
 
 package org.elasticsearch.search;
 
-import org.apache.lucene.tests.util.LuceneTestCase;
 import org.elasticsearch.ExceptionsHelper;
 import org.elasticsearch.action.ActionFuture;
 import org.elasticsearch.action.search.MultiSearchResponse;
@@ -50,8 +49,7 @@ import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.hasSize;
 import static org.hamcrest.Matchers.notNullValue;
 
-@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.SUITE)
-@LuceneTestCase.AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/102257")
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST)
 public class SearchCancellationIT extends AbstractSearchCancellationTestCase {
 
     @Override
@@ -288,12 +286,11 @@ public class SearchCancellationIT extends AbstractSearchCancellationTestCase {
                     assertTrue("All SearchShardTasks should then be cancelled", shardQueryTask.isCancelled());
                 }
             }, 30, TimeUnit.SECONDS);
-            shardTaskLatch.countDown(); // unblock the shardTasks, allowing the test to conclude.
         } finally {
+            shardTaskLatch.countDown(); // unblock the shardTasks, allowing the test to conclude.
             searchThread.join();
-            for (ScriptedBlockPlugin plugin : plugins) {
-                plugin.setBeforeExecution(() -> {});
-            }
+            plugins.forEach(plugin -> plugin.setBeforeExecution(() -> {}));
+            searchShardBlockingPlugins.forEach(plugin -> plugin.setRunOnNewReaderContext((ReaderContext c) -> {}));
         }
     }
 

--- a/test/framework/src/main/java/org/elasticsearch/test/AbstractSearchCancellationTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/AbstractSearchCancellationTestCase.java
@@ -210,7 +210,7 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             }
         }
 
-        public void waitForBlock(int timeout, TimeUnit timeUnit) {
+        public void waitForLock(int timeout, TimeUnit timeUnit) {
             try {
                 assertTrue(shouldBlock.tryAcquire(timeout, timeUnit));
                 shouldBlock.release(1);
@@ -227,7 +227,7 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             LeafStoredFieldsLookup fieldsLookup = (LeafStoredFieldsLookup) params.get("_fields");
             LogManager.getLogger(AbstractSearchCancellationTestCase.class).info("Blocking on the document {}", fieldsLookup.get("_id"));
             hits.incrementAndGet();
-            waitForBlock(10, TimeUnit.SECONDS);
+            waitForLock(10, TimeUnit.SECONDS);
             return true;
         }
 
@@ -247,7 +247,7 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             }
             logIfBlocked("Blocking in reduce");
             hits.incrementAndGet();
-            waitForBlock(10, TimeUnit.SECONDS);
+            waitForLock(10, TimeUnit.SECONDS);
             return 42;
         }
 
@@ -258,7 +258,7 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             }
             logIfBlocked("Blocking in map");
             hits.incrementAndGet();
-            waitForBlock(10, TimeUnit.SECONDS);
+            waitForLock(10, TimeUnit.SECONDS);
             return 1;
         }
 

--- a/test/framework/src/main/java/org/elasticsearch/test/AbstractSearchCancellationTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/AbstractSearchCancellationTestCase.java
@@ -35,7 +35,8 @@ import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
-import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
@@ -152,7 +153,7 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
 
         private final AtomicInteger hits = new AtomicInteger();
 
-        private final AtomicBoolean shouldBlock = new AtomicBoolean(true);
+        private final Semaphore shouldBlock = new Semaphore(Integer.MAX_VALUE);
 
         private final AtomicReference<Runnable> beforeExecution = new AtomicReference<>();
 
@@ -161,11 +162,16 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
         }
 
         public void disableBlock() {
-            shouldBlock.set(false);
+            shouldBlock.release(Integer.MAX_VALUE);
         }
 
         public void enableBlock() {
-            shouldBlock.set(true);
+            try {
+                shouldBlock.acquire(Integer.MAX_VALUE);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new AssertionError(e);
+            }
         }
 
         public void setBeforeExecution(Runnable runnable) {
@@ -196,6 +202,23 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             );
         }
 
+        public void tryBlock(String logMessage) {
+            if (shouldBlock.tryAcquire(1) == false) {
+                LogManager.getLogger(AbstractSearchCancellationTestCase.class).info(logMessage);
+            } else {
+                shouldBlock.release(1);
+            }
+        }
+
+        public void waitForBlock(int timeout, TimeUnit timeUnit) {
+            try {
+                assertTrue(shouldBlock.tryAcquire(timeout, timeUnit));
+                shouldBlock.release(1);
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+
         private Object searchBlockScript(Map<String, Object> params) {
             final Runnable runnable = beforeExecution.get();
             if (runnable != null) {
@@ -204,11 +227,7 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             LeafStoredFieldsLookup fieldsLookup = (LeafStoredFieldsLookup) params.get("_fields");
             LogManager.getLogger(AbstractSearchCancellationTestCase.class).info("Blocking on the document {}", fieldsLookup.get("_id"));
             hits.incrementAndGet();
-            try {
-                assertBusy(() -> assertFalse(shouldBlock.get()));
-            } catch (Exception e) {
-                throw new RuntimeException(e);
-            }
+            waitForBlock(10, TimeUnit.SECONDS);
             return true;
         }
 
@@ -226,15 +245,9 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             if (runnable != null) {
                 runnable.run();
             }
-            if (shouldBlock.get()) {
-                LogManager.getLogger(AbstractSearchCancellationTestCase.class).info("Blocking in reduce");
-            }
+            tryBlock("Blocking in reduce");
             hits.incrementAndGet();
-            try {
-                assertBusy(() -> assertFalse(shouldBlock.get()));
-            } catch (Exception e) {
-                throw new RuntimeException(e);
-            }
+            waitForBlock(10, TimeUnit.SECONDS);
             return 42;
         }
 
@@ -243,15 +256,9 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             if (runnable != null) {
                 runnable.run();
             }
-            if (shouldBlock.get()) {
-                LogManager.getLogger(AbstractSearchCancellationTestCase.class).info("Blocking in map");
-            }
+            tryBlock("Blocking in map");
             hits.incrementAndGet();
-            try {
-                assertBusy(() -> assertFalse(shouldBlock.get()));
-            } catch (Exception e) {
-                throw new RuntimeException(e);
-            }
+            waitForBlock(10, TimeUnit.SECONDS);
             return 1;
         }
 

--- a/test/framework/src/main/java/org/elasticsearch/test/AbstractSearchCancellationTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/AbstractSearchCancellationTestCase.java
@@ -202,7 +202,7 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             );
         }
 
-        public void tryBlock(String logMessage) {
+        public void tryBlockOnce(String logMessage) {
             if (shouldBlock.tryAcquire(1) == false) {
                 LogManager.getLogger(AbstractSearchCancellationTestCase.class).info(logMessage);
             } else {
@@ -245,7 +245,7 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             if (runnable != null) {
                 runnable.run();
             }
-            tryBlock("Blocking in reduce");
+            tryBlockOnce("Blocking in reduce");
             hits.incrementAndGet();
             waitForBlock(10, TimeUnit.SECONDS);
             return 42;
@@ -256,7 +256,7 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             if (runnable != null) {
                 runnable.run();
             }
-            tryBlock("Blocking in map");
+            tryBlockOnce("Blocking in map");
             hits.incrementAndGet();
             waitForBlock(10, TimeUnit.SECONDS);
             return 1;

--- a/test/framework/src/main/java/org/elasticsearch/test/AbstractSearchCancellationTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/AbstractSearchCancellationTestCase.java
@@ -202,7 +202,7 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             );
         }
 
-        public void tryBlockOnce(String logMessage) {
+        public void logIfBlocked(String logMessage) {
             if (shouldBlock.tryAcquire(1) == false) {
                 LogManager.getLogger(AbstractSearchCancellationTestCase.class).info(logMessage);
             } else {
@@ -245,7 +245,7 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             if (runnable != null) {
                 runnable.run();
             }
-            tryBlockOnce("Blocking in reduce");
+            logIfBlocked("Blocking in reduce");
             hits.incrementAndGet();
             waitForBlock(10, TimeUnit.SECONDS);
             return 42;
@@ -256,7 +256,7 @@ public class AbstractSearchCancellationTestCase extends ESIntegTestCase {
             if (runnable != null) {
                 runnable.run();
             }
-            tryBlockOnce("Blocking in map");
+            logIfBlocked("Blocking in map");
             hits.incrementAndGet();
             waitForBlock(10, TimeUnit.SECONDS);
             return 1;

